lighthouse-core/audits/consistently-interactive.js

```diff
@@ -0,0 +1,248 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Audit = require('./audit');
+const TracingProcessor = require('../lib/traces/tracing-processor');
+const Formatter = require('../report/formatter');
+
+// Parameters (in ms) for log-normal CDF scoring. To see the curve:
+//   https://www.desmos.com/calculator/uti67afozh
+const SCORING_POINT_OF_DIMINISHING_RETURNS = 1700;
+const SCORING_MEDIAN = 10000;
+const SCORING_TARGET = 5000;
+
+const REQUIRED_QUIET_WINDOW = 5000;
+const ALLOWED_CONCURRENT_REQUESTS = 2;
+
+const distribution = TracingProcessor.getLogNormalDistribution(
+  SCORING_MEDIAN,
+  SCORING_POINT_OF_DIMINISHING_RETURNS
+);
+
+/**
+ * @fileoverview This audit identifies the time the page is "consistently interactive".
+ * Looks for the first period of at least 5 seconds after FMP where both CPU and network were quiet,
+ * and returns the timestamp of the beginning of the CPU quiet period.
+ * @see https://docs.google.com/document/d/1GGiI9-7KeY3TPqS3YT271upUVimo-XiL5mwWorDUD4c/edit#
+ */
+class ConsistentlyInteractiveMetric extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Performance',
+      name: 'consistently-interactive',
+      description: 'Consistently Interactive (beta)',
+      helpText: 'The point at which most network resources have finished loading and the ' +
+          'CPU is idle for a prolonged period.',
+      optimalValue: SCORING_TARGET.toLocaleString() + 'ms',
+      scoringMode: Audit.SCORING_MODES.NUMERIC,
+      requiredArtifacts: ['traces', 'devtoolsLogs']
+    };
+  }
+
+  /**
+   * @param {!Array} networkRecords
+   * @param {{timestamps: {traceEnd: number}}} traceOfTab
+   * @return {!Array<{start: number, end: number}>}
+   */
+  static _findNetworkQuietPeriods(networkRecords, traceOfTab) {
+    const traceEnd = traceOfTab.timestamps.traceEnd;
+
+    // First collect the timestamps of when requests start and end
+    const timeBoundaries = [];
+    networkRecords.forEach(record => {
+      const scheme = record.parsedURL && record.parsedURL.scheme;
+      if (scheme === 'data' || scheme === 'ws') {
```

Comment:
might be overkill, but should we convert this to a `ignoredNetworkSchemes` array and pull it into a constant?